### PR TITLE
Fix #2277, #682: Introduce utility for framework agnostic activity rotation

### LINKED ISSUE #682: Refactor previous code for testing orientation changes with the custom function written for it [Blocked on Landscape mode]
There are a lot of test files that use the espresso methods directly to change orientation while testing the app, even though a custom method exists for the same. 
Basically there are different types of orientation change related test cases, which have bee implemented in different ways. But we need all of them to be consistent and all of them should use only `orientationLandscape()` function to do orientation change.

The file that needs refactoring are:

- [x] `TopicPracticeFragmentTest.kt`
- [ ] `InputInteractionViewTestActivity.kt`
- [x] `StateFragmentTest.kt`
- [x] `AudioFragmentTest.kt`
- [x] `ConceptCardFragmentTest.kt`
- [x] `TopicLessonsFragmentTest.kt`
- [x] `TopicReviewFragmentTest.kt`
- [x] `TopicInfoFragmentTest.kt`

### LINKED ISSUE #2277: [Investigate] Does Robolectric performing Orientation change correctly
**Describe the bug**
Orientation change is not behaving correctly in robolectric test cases. 

**Expected behavior**
Need to investigate this and see if can have a better approach or modifying the current one.

### PR DESCRIPTION
**Pull request by BenHenning**
Fix #2277
Fix #682

This PR replaces our existing action for managing configuration changes with a new utility that can perform proper activity rotations with both configuration changes & activity recreation for both Robolectric & Espresso tests.

Some specifics:
- This uses RuntimeEnvironment.setQualifiers() to alter the orientation + an activity recreation (via ActivityScenario) to trigger the destroy/recreate flows
- This introduces a new injector/provider to allow us to use a very nice ActivityScenario extension syntax (versus needing to inject the rotator utility & pass the scenario into the methods)

### CODE REVIEW

**Comment by BenHenning** [comment]
@anandwana001 & @FareesHussain what do you think about this? It doesn't yet have tests, but I did try to update all existing tests to use the utility. I was even able to update one test that was previously failing in Robolectric but now passes (in TopicInfoFragmentTest).

**Review by FareesHussain — CHANGES_REQUESTED** [review]
Error needs to be resolved.
https://github.com/oppia/oppia-android/blob/e15c6a71a20d468801772612ff1351440580cec2/app/src/sharedTest/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivityTest.kt#L236

I tried to run the test (AdministratorActivityTest) in both Espresso and Robolectric
From what I've observed this implementation works for Espresso but Yet it does'nt work for robolectric

```kotlin
// This test needs to fail but it passes for robolectric in this implementation too
  @Test
  fun testAdministratorControlsFragment_configChange_clickLogout_logoutDialogIsDisplayed_rotateLandscape_dialogDisppears() {
    launch<AdministratorControlsActivity>(
      createAdministratorControlsActivityIntent(
        profileId = 0
      )
    ).use {
      testCoroutineDispatchers.runCurrent()
      scrollToPosition(position = 4)
      onView(withId(R.id.log_out_text_view)).perform(click())
      verifyTextInDialog(textInDialogId = R.string.log_out_dialog_message)
      verifyTextInDialog(textInDialogId = R.string.log_out_dialog_okay_button)
      verifyTextInDialog(textInDialogId = R.string.log_out_dialog_cancel_button)
      it.rotateToLandscape()
      verifyTextInDialog(textInDialogId = R.string.log_out_dialog_message)
      verifyTextInDialog(textInDialogId = R.string.log_out_dialog_okay_button)
      verifyTextInDialog(textInDialogId = R.string.log_out_dialog_cancel_button)
    }
  }
```

I'm not sure why this test case fails for this implementation
**AdministratorControlsActivityTest -> testAdministratorControlsFragment_configChange_clickLogout_logoutDialogIsDisplayed**
```
androidx.test.espresso.NoMatchingRootException: Matcher 'is dialog' did not match any of the following roots: [Root{application-window-token=android.view.ViewRootImpl$W@7d1e4aad, window-token=android.view.ViewRootImpl$W@7d1e4aad, has-window-focus=true, layout-params-type=1, layout-params-string={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x7f110005
  fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH HARDWARE_ACCELERATED DRAWS_SYSTEM_BAR_BACKGROUNDS
  pfl=FORCE_DRAW_STATUS_BAR_BACKGROUND}, decor-view-string=DecorView{id=-1, visibility=VISIBLE, width=960, height=1410, has-focus=true, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x7f110005
  fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH HARDWARE_ACCELERATED DRAWS_SYSTEM_BAR_BACKGROUNDS
  pfl=FORCE_DRAW_STATUS_BAR_BACKGROUND}, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=1}}, Root{application-window-token=android.view.ViewRootImpl$W@2d543180, window-token=android.view.ViewRootImpl$W@2d543180, has-window-focus=true, layout-params-type=1, layout-params-string={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x7f110005
  fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH HARDWARE_ACCELERATED DRAWS_SYSTEM_BAR_BACKGROUNDS
  pfl=FORCE_DRAW_STATUS_BAR_BACKGROUND}, decor-view-string=DecorView{id=-1, visibility=VISIBLE, width=960, height=1410, has-focus=true, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x7f110005
  fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH HARDWARE_ACCELERATED DRAWS_SYSTEM_BAR_BACKGROUNDS
  pfl=FORCE_DRAW_STATUS_BAR_BACKGROUND}, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=1}}]

	at java.lang.Thread.getStackTrace(Thread.java:1559)
	at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:96)
	at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:59)
	at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:322)
	at androidx.test.espresso.ViewInteraction.check(ViewInteraction.java:306)
	at org.oppia.android.app.administratorcontrols.AdministratorControlsActivityTest.verifyTextInDialog(AdministratorControlsActivityTest.kt:562)
	at org.oppia.android.app.administratorcontrols.AdministratorControlsActivityTest.testAdministratorControlsFragment_configChange_clickLogout_logoutDialogIsDisplayed(AdministratorControlsActivityTest.kt:347)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:546)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:252)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by BenHenning** [comment]
Reopening to trigger actions (since they didn't run earlier for some reason).

**Comment by BenHenning** [comment]
I've run into significant issues with a bunch of tests when trying to run all of app module locally. In particular:
- Some tests are failing due to (presumably) issues in the test or scenario now that rotation is actually happening
- Some tests are hanging permanently. One main case I identified in AddProfileActivityTest was due to an infinite measure being triggered by TextInputLayout--fortunately that particular test could be removed because it was testing an implementation behavior of TextInputLayout. However, there are other tests hanging and they may be due to similar issues.
- There seems to be some weird situation where certain views aren't being scrolled to or measured correctly when rotated TO landscape mode. I'm not sure why this is happening, and digging into it for a while didn't get me very far. All the dimensions are believable, it just isn't horizontally aligned correctly. This only happens after rotation.

It's not surprising a lot of the tests are having issues, but this seems like it's going to be a significant effort to solve.

Instead, my plan is to:
1. Introduce a general-purpose "changeConfiguration" function that triggers a recreate with a non-rotation configuration property so that we can test configuration changes without needing to worry about orientation changes.
2. Update all tests that don't care about orientation changes to use the new change configuration flow.
3. Fix failures from tests updated in (2).
4. Verify the tests from (2) also pass on Espresso (since orientation change will now no longer be used for this purpose)
5. Introduce an optional parameter in the new rotation functions to disable the actual rotation part for Robolectric
6. Only enable the flag from (5) in tests that need it.
7. For each test for (6), add TODOs that point to a new issue to track fixing orientation changes in Robolectric tests
8. Send this PR for full review & check it in

Long-term we'll need to fix the issue filed in (7), and I suspect this will be a fairly challenging set of problems. We shouldn't prioritize that work since the tests will still provide some value (especially when run in Espresso).

**Comment by anandwana001** [comment]
The plan mentioned above sounds good. Will check this PR once ready for review. Keeping this pr pinned to look at later today.

**Comment by anandwana001** [comment]
As we are moving towards a new way to change orientation, can we close this issue - https://github.com/oppia/oppia-android/issues/682?

**Comment by BenHenning** [comment]
This seems like a reasonable fix for #682. Updating it to specify that, too.

File: testing/src/main/java/org/oppia/android/testing/ActivityRotator.kt [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+      retrieveActivityRotator().changeOrientation(
+        scenario = this,
+        newOrientation = Orientation.LANDSCAPE
+      )
+
+    private fun retrieveActivityRotator(): ActivityRotator {
+      val context = ApplicationProvider.getApplicationContext<Context>()
+      check(context is Provider) {
+        "Error: the TestApplication is not implementing ActivityRotator.Provider. Please update " +
+          "he test component before using the companion rotateToLandscape/rotateToPortrait " +
```
Comment (FareesHussain) [inline testing/src/main/java/org/oppia/android/testing/ActivityRotator.kt:101]: NIT fix
```suggestion
          "the test component before using the companion rotateToLandscape/rotateToPortrait " +
```

**Comment by BenHenning** [comment]
Per the problems in https://github.com/oppia/oppia-android/pull/2572#issuecomment-770158899 & lack of time, I don't think I'll be able to get to this for a while so closing & adding to my backlog for now.

**Pull request closed without merging**